renku/core/metadata/gateway/database_gateway.py

```diff
@@ -17,110 +17,123 @@
+):
+    """Dump relation entry to database."""
+    btree = database_dispatcher.current_database["_downstream_relations"]
 
+    btree[relation.id] = relation
 
-@inject.autoparams()
-def upstream_activity(activity: Activity, catalog, database: Database) -> Iterator[Activity]:
-    """Map an activity to its upstream predecessors."""
-    result = []
-    for usage in activity.usages:
-        if not isinstance(usage.entity, Collection):
-            # NOTE: Get direct dependants
-            result.extend(database["activities-by-generation"].get(usage.entity.path, []))
-        else:
-            # NOTE: Get dependants that are in a generated directory
-            for path, activities in database["activities-by-generation"].items():
-                parent = Path(usage.entity.path).resolve()
-                child = Path(path).resolve()
-                if parent == child or parent in child.parents:
-                    result.extend(activities)
-
-    return result
+    return relation.id
+
+
+def load_downstream_relations(token, catalog, cache, database_dispatcher: IDatabaseDispatcher):
+    """Load relation entry from database."""
+    btree = database_dispatcher.current_database["_downstream_relations"]
+
+    return btree[token]
 
 
 # NOTE: Transitive query factory is needed for transitive (follow more than 1 edge) queries
-downstream_transitive_factory = TransposingTransitive(RELATION, "downstream_activity")
-upstream_transitive_factory = TransposingTransitive(RELATION, "upstream_activity")
+downstream_transitive_factory = TransposingTransitive("downstream", "upstream")
 
 
 class DatabaseGateway(IDatabaseGateway):
     """Gateway for base database operations."""
 
-    database = inject.attr(Database)
+    database_dispatcher = inject.attr(IDatabaseDispatcher)
 
     def initialize(self) -> None:
         """Initialize the database."""
-        self.database.clear()
+        database = self.database_dispatcher.current_database
+
+        database.clear()
 
-        self.database.add_index(name="activities", object_type=Activity, attribute="id")
-        self.database.add_index(name="latest-activity-by-plan", object_type=Activity, attribute="association.plan.id")
-        self.database.add_root_object(name="activities-by-usage", obj=BTrees.OOBTree.OOBTree())
-        self.database.add_root_object(name="activities-by-generation", obj=BTrees.OOBTree.OOBTree())
+        database.add_index(name="activities", object_type=Activity, attribute="id")
+        database.add_index(name="latest-activity-by-plan", object_type=Activity, attribute="association.plan.id")
+        database.add_root_object(name="activities-by-usage", obj=BTrees.OOBTree.OOBTree())
+        database.add_root_object(name="activities-by-generation", obj=BTrees.OOBTree.OOBTree())
 
-        activity_catalog = Catalog(dump_activity, load_activity, btree=BTrees.family32.OO)
+        database.add_root_object(name="_downstream_relations", obj=BTrees.OOBTree.OOBTree())
```

Comment:
just because this is something added by the zc.relation.Catalog, not something someone should ever manually edit/access. So it's a "private" object, just following python conventions.